### PR TITLE
Ports SecurEye a camera app for modular pcs from /tg/

### PR DESCRIPTION
**Pull request by BrumalLake**
## About The Pull Request
Ports [tgstation#54225](https://github.com/tgstation/tgstation/pull/54225)
SecurEye is an app which people with security access can download to view cameras
It can only be downloaded on laptops or computers
The camera console search bar also follows if you scroll down the list now

## Why It's Good For The Game
Makes laptops better while still making you have to carry them in one hand to see cameras as well as an on the move option for members of security who bother to get a laptop

## Changelog
:cl: zxaber, lordScrubling
add: Adds a camera viewing app, SecurEye, available for download on laptops and modular consoles to those that have security access.
tweak: Camera console UIs (including for SecurEye) now have the search field locked on screen, rather than scrolling up with the camera list.
/:cl:

### CODE REVIEW

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		// Ghosts shouldn't count towards concurrent users, which produces
+		// an audible terminal_on click.
+		if(is_living)
+			concurrent_users += user_ref
+		// Register map objects
+		user.client.register_map_obj(cam_screen)
+		for(var/plane in cam_plane_masters)
+			user.client.register_map_obj(plane)
+		user.client.register_map_obj(cam_background)
+		return ..()
```
Comment (utilbird) [inline code/modules/modular_computers/file_system/programs/secureye.dm]:
```suggestion
	if(ui)
		return
	var/user_ref = REF(user)
	var/is_living = isliving(user)
	// Ghosts shouldn't count towards concurrent users, which produces
	// an audible terminal_on click.
	if(is_living)
		concurrent_users += user_ref
	// Register map objects
	user.client.register_map_obj(cam_screen)
	for(var/plane in cam_plane_masters)
		user.client.register_map_obj(plane)
	user.client.register_map_obj(cam_background)
	return ..()
```

Comment (BrumalLake) [reply code/modules/modular_computers/file_system/programs/secureye.dm]: I think it made more sense before

**Comment by Crossedfall** [comment]
Because this was rushed through, there's no message from the action stating that it doesn't have permission to push to your repo. 

> [securEye 8af92b7] TGUI Rebuild
 1 file changed, 0 insertions(+), 0 deletions(-)
 rewrite tgui/public/tgui.bundle.js (82%)
To https://github.com/lordScrubling/junk
 ! [remote rejected] securEye -> securEye (permission denied)
error: failed to push some refs to 'https://github.com/lordScrubling/junk'
Error: Process completed with exit code 1.

You need to ensure maintainers are allowed to push edits. (Honestly count yourself lucky on that one)

**Comment by BrumalLake** [comment]
/rebuild tgui
just thought about that yeah

**Comment by BrumalLake** [comment]
?rebuild tgui

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	var/list/concurrent_users = list()
+
+	// Stuff needed to render the map
+	var/map_name
+	var/atom/movable/screen/map_view/cam_screen
+	/// All the plane masters that need to be applied.
+	var/list/cam_plane_masters
+	var/atom/movable/screen/background/cam_background
+
+/datum/computer_file/program/secureye/New()
```
Comment (Archanial) [inline code/modules/modular_computers/file_system/programs/secureye.dm:27]: this shouldn't override `new()`, put this in `initialize()`

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		)
+	return data
+
+/datum/computer_file/program/secureye/ui_static_data()
+	var/list/data = list()
+	data["mapRef"] = map_name
+	var/list/cameras = get_available_cameras()
+	data["cameras"] = list()
+	for(var/i in cameras)
+		var/obj/machinery/camera/C = cameras[i]
```
Comment (Archanial) [inline code/modules/modular_computers/file_system/programs/secureye.dm:96]: Can anything else be other than `machinery/camera` be in cameras? If not, use typed loop with `as()`.

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+/datum/computer_file/program/secureye/ui_close(mob/user)
+	. = ..()
+	var/user_ref = REF(user)
+	var/is_living = isliving(user)
+	// Living creature or not, we remove you anyway.
+	concurrent_users -= user_ref
+	// Unregister map objects
+	user.client.clear_map(map_name)
+	// Turn off the console
+	if(length(concurrent_users) == 0 && is_living)
```
Comment (Archanial) [inline code/modules/modular_computers/file_system/programs/secureye.dm:132]:
```suggestion
	if(!length(concurrent_users) && is_living)
```
what is `is_living` for here?

**Review by Archanial — DISMISSED** [review]

File: code/modules/modular_computers/file_system/programs/secureye.dm (hunk 3 above)
Comment (BrumalLake) [reply code/modules/modular_computers/file_system/programs/secureye.dm:96]: borgs, i believe

File: code/modules/modular_computers/file_system/programs/secureye.dm (hunk 4 above)
Comment (BrumalLake) [reply code/modules/modular_computers/file_system/programs/secureye.dm:132]: no clue, but the line mirrors camera.dm so i'm not in to change it

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Tested in our staging environment. Doesn't seem to work anymore, there might have been a change since this was opened that broke it.

**Comment by BrumalLake** [comment]
works perfectly again, thought uis updated automatically by default instead of the other way around
not sure why this is in camera console
image

**Comment by Archanial** [comment]
*sigh
image

**Comment by lordScrubling** [comment]
https://discord.com/channels/427337870844362753/565330972707651584/916517063206522910

**Comment by BrumalLake** [comment]
> *sigh image

https://discord.com/channels/427337870844362753/565330972707651584/916517063206522910

**Review by KubeRoot — DISMISSED** [review]
Tested locally, works for me. If we want to be extra sure, recommend testmerging again.

**Review by Crossedfall — APPROVED** [review]
No issues during TM. Had a warden test it out for me as well.

**Pull request merged**